[move-prover] Prover tool installation scripts and CI integration

wrwg opened the pull request:
This PR integrates the installation logic for the Move prover tools in the general libra script at `./scripts/dev-setup.sh`. This script is also used for CI, enabling prover tests to run as part of regular continuous integration testing.

Two new flags are added to `dev-setup.sh`:

- `dev-setup.sh -z` installs or updates only the prover tools.
- `dev-setup.sh -y <other flags>` installs the full development environment including the prover tools (and what else is selected with `<other flags>`.

The move-prover documentation has been updated to describe the new installation procedure. With this PR landing, Move prover tests are active in CI.

This PR also adds general support for `dnf` as a package manager in the dev-setup script. This allows f.i. to run the script on a FB devserver. This support is still experimental and has not been deeply tested yet.


## Motivation

Full verification testing in CI and better tool installation support/documentation.

Closes  #3935

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

NA

## Related PRs

NA

bmwill commented:
This is awesome! I'm excited that we can finally get these tests running in CI. I'm going to also add @rexhoffman to review the dev-setup.sh part as he most recently spent some time to overhaul it.

rexhoffman commented on scripts/dev_setup.sh:
```diff
@@ -27,18 +30,30 @@ function usage {
-  echo "-o intall operations tooling as well: helm, terraform, hadolint, yamllint, vault, docker, kubectl, python3"
+  echo "-o install operations tooling as well: helm, terraform, hadolint, yamllint, vault, docker, kubectl, python3"
+  echo "-y install Move prover tools: z3, dotnet, boogie"
+  echo "-z installs or updates Move prover tools only"
```
how about switching this out to only install the validator/rust tools if -v is passed in?   would be cleaner to be fully additive.

wrwg replied:
I agree but leave that for another PR. I'm not sure whether anybody depends on the current semantics of the flags.

rexhoffman replied:
only place depending on the flags is the tests in docker/ci/....

wrwg replied:
Ok, done this now. There is a new option `-t` ("tools") for installing the build tools. I did a search over the project, and there where quite a few occurrences also in documentation. I hope we don't have any docs outside of the repo which refer to this.

rexhoffman commented on scripts/dev_setup.sh:
```diff
@@ -27,18 +30,30 @@ function usage {
   echo "should be called from the root folder of the libra project"
 }
 
+function add_to_profile {
```
good addition!

rexhoffman commented on scripts/dev_setup.sh:
```diff
@@ -27,18 +30,30 @@ function usage {
 function update_path_and_profile {
   touch "${HOME}"/.profile
   mkdir -p "${HOME}"/bin
-  export PATH="${HOME}"/bin:"${HOME}"/.cargo/bin:"${PATH}"
```
I _think_ the path needed updated in this shell for rust tooling to install on a fresh system, but the pr tests should show us this....

rexhoffman replied:
Scratch that, the tests are passing.

rexhoffman reviewed (commented):
So this file is tested in circleci on update -- you can verify your installation script works in several systems easily if you like, instructions below.....

Just add a check to ensure that the tools are all installed on the base images in docker/ci if you want them in ci, and to test your script on various systems. 

This one in particular will make sure your tools are available in CI if you want it here...

https://github.com/libra/libra/blob/master/docker/ci/circleci/Dockerfile#L20-L32

And ugh, dnf....  we can support testing that with a centos 8 image build I think.  I think I dodged this while I was building the updater.....

rexhoffman reviewed (approved):
Regardless I'm good with merging this and I can pick up the cleanups if necessary.

wrwg commented:
> So this file is tested in circleci on update -- you can verify your installation script works in several systems easily if you like, instructions below.....

Cool, done -- added the tests.

wrwg commented:
Thanks to @rexhoffman for pointing to the awesome support for testing build on multiple linux platforms! I have the prover tool install now working on all platforms we are testing for (not just ubuntu as we are using in CI). 

Also changed the script now based on reviewer feedback so all components of installations can be optional. However, to avoid any change to existing use and documentation (`./scripts/dev-setup.sh` is mentioned in various getting started docs),  the default is if setup is called with no options, it's like being called with the new option `-t` which guards installation of basic build tools and rust.

As regards the added support for `dnf`, I'm not sure whether it's needed. The only reason I added it because on CentOS as used on FB devservers, `yum` is not available by default. However, if I launch a standard CentOS docker image, `yum` is there, and in fact `dnf` can install it if wanted also on devservers. Not really sure what's going on here but leaving it in for others to further investigate.

Landing this now. Thanks for the reviews.

wrwg commented:
/land
